# --- Day 20: Firewall Rules ---

You'd like to set up a small hidden computer here so you can use it to get back into the network later. However, the corporate firewall only allows communication with certain external IP addresses.

You've retrieved the list of blocked IPs from the firewall, but the list seems to be messy and poorly maintained, and it's not clear which IPs are allowed. Also, rather than being written in dot-decimal notation, they are written as plain 32-bit integers, which can have any value from 0 through 4294967295, inclusive.

For example, suppose only the values 0 through 9 were valid, and that you retrieved the following blacklist:

```
5-8
0-2
4-7```

The blacklist specifies ranges of IPs (inclusive of both the start and end value) that are not allowed. Then, the only IPs that this firewall allows are 3 and 9, since those are the only numbers not in any range.

**Given the list of blocked IPs you retrieved from the firewall (your puzzle input), what is the lowest-valued IP that is not blocked?**

In [1]:
with open(f'inputs/20.txt') as f:
    data = f.read().strip().split()
data[:3]

['272152717-281364173', '480675455-489214207', '3562619188-3566180938']

first up, parsing the data:

In [2]:
blocklist = []
for line in data:
    r = [int(i) for i in line.split("-")]
    blocklist.append(r)
blocklist[:3]

[[272152717, 281364173], [480675455, 489214207], [3562619188, 3566180938]]

In [47]:
test_list = [[5,8],[0,2],[4,7]]
test_list

[[5, 8], [0, 2], [4, 7]]

Now to sort the blocklists to make it easy to look up if a number is in the blocklist:

In [48]:
blocklist = sorted(blocklist, key=lambda x: (x[0], x[1]))
test_list = sorted(test_list, key=lambda x: (x[0], x[1]))
blocklist[:4]

[[0, 166475], [110217, 574651], [574652, 1770165], [1770166, 1994691]]

so trying a simple way where I go through each port pair until I find the first unblocked one:

In [87]:
def find_first_unblocked_port(blocklist=test_list):
    low_num = blocklist[0][0]
    max_num = blocklist[0][1]
    
    if low_num > 0:
        return 0
    
    for lower, upper in blocklist[1:]:
        if lower > max_num + 1:
            print(f"Found it! {max_num+1:,} is the first unblocked port.")
            return max_num + 1
        elif upper > max_num:
                max_num = upper 
    return False

find_first_unblocked_port()

Found it! 3 is the first unblocked port.


3

In [88]:
%time find_first_unblocked_port(blocklist)

Found it! 23,923,783 is the first unblocked port.
CPU times: user 2 ms, sys: 0 ns, total: 2 ms
Wall time: 2.21 ms


23923783

`23923783` is the answer to part 1

# --- Part Two ---

**How many IPs are allowed by the blacklist?**

Here, I'm just modifying the above func to go through the entire blocklist and count the unblocked ports:

In [90]:
def total_unblocked_ports(blocklist=test_list):
    low_num = blocklist[0][0]
    max_num = blocklist[0][1]
    
    unblocked_ports = low_num

    for lower, upper in blocklist[1:]:
        if lower > max_num + 1:
            unblocked_ports += lower - max_num - 1
            low_num = lower
            max_num = upper
        
        elif upper > max_num:
               max_num = upper
    
    return unblocked_ports

%time total_unblocked_ports(blocklist)

CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 239 µs


125

`125` is the ans for part 2.